### PR TITLE
[Improvement][Master] Add Misfire policy control in schedule

### LINKED ISSUE #15986: [Improvement][Master] Add Misfire policy control in schedule
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

Now our trigger are configured in fixed misfire instruction.
We can let each cronjob to choose theirselves.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by pegasas**
## Purpose of the pull request

fix: https://github.com/apache/dolphinscheduler/issues/15986

## Brief change log

Now our trigger are configured in fixed misfire instruction.
We can let each cronjob to choose theirselves.

## Verify this pull request

image

### CODE REVIEW

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/MisfirePolicy.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    private final String descp;
+
+    public int getCode() {
+        return code;
+    }
+
+    public String getDescp() {
+        return descp;
+    }
+}
```
Comment (ruanwenjun) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/MisfirePolicy.java]: The misfire policy is a concept in DS, quartz is just a implementation, we will have other implementations in the future, don't keep it same with quartz.

```suggestion
public enum MisfirePolicy {

    EXACTLY_ONCE("All misfired schedule trigger will be handle"),
    KEEP_LATEST("Only handle the latest misfired schedule trigger"),
    DROP_MISFIRED("Will drop the misfired schedule trigger"),
    ;

    MisfirePolicy(int code, String descp) {
        this.code = code;
        this.descp = descp;
    }

    @EnumValue
    private final int code;
    private final String descp;

    public int getCode() {
        return code;
    }

    public String getDescp() {
        return descp;
    }
}
```

File: dolphinscheduler-dao/src/main/resources/sql/upgrade/3.2.2_schema/mysql/dolphinscheduler_ddl.sql [hunk 2]
```diff
@@ -53,4 +53,6 @@ END;
 d//
 delimiter ;
 CALL modify_data_t_ds_audit_log_input_entry;
-DROP PROCEDURE modify_data_t_ds_audit_log_input_entry;
\ No newline at end of file
+DROP PROCEDURE modify_data_t_ds_audit_log_input_entry;
+
+ALTER TABLE t_ds_schedules ADD COLUMN misfire_policy tinyint(4) NOT NULL;
```
Comment (ruanwenjun) [inline dolphinscheduler-dao/src/main/resources/sql/upgrade/3.2.2_schema/mysql/dolphinscheduler_ddl.sql]: Need to refresh the history data.

File: dolphinscheduler-scheduler-plugin/dolphinscheduler-scheduler-quartz/src/main/java/org/apache/dolphinscheduler/scheduler/quartz/QuartzCornTriggerBuilder.java [hunk 3]
```diff
@@ -81,14 +81,28 @@ public CronTrigger build() {
...
+                break;
+            case FIREANDPROCEED:
+                cronScheduleBuilder.withMisfireHandlingInstructionFireAndProceed();
+                break;
+            case IGNOREMISFIRES:
+            default:
+                cronScheduleBuilder.withMisfireHandlingInstructionIgnoreMisfires();
+                break;
+        }
+
```
Comment (ruanwenjun) [inline dolphinscheduler-scheduler-plugin/dolphinscheduler-scheduler-quartz/src/main/java/org/apache/dolphinscheduler/scheduler/quartz/QuartzCornTriggerBuilder.java]: Is this PR finished? how can we set this property?

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/MisfirePolicy.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * misfire policy when using quartz scheduler
+ */
+public enum MisfirePolicy {
+
+    /**
+     * 0 ending process when some tasks failed.
+     * 1 continue running when some tasks failed.
+     **/
+    IGNOREMISFIRES(-1, "ignoremisfires"),
```
Comment (rickchengx) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/MisfirePolicy.java]: May I ask Why does `code` starts from -1?

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/MisfirePolicy.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+package org.apache.dolphinscheduler.common.enums;
+
+import com.baomidou.mybatisplus.annotation.EnumValue;
+
+/**
+ * misfire policy when using quartz scheduler
+ */
+public enum MisfirePolicy {
```
Comment (rickchengx) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/MisfirePolicy.java:25]: What is the granularity of this policy, server level or scheduling level? If it is a scheduling level, is a corresponding front end needed to allow users to set it?

Also, doc needs to be updated

**Comment by qingwli** [comment]
Mark

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/MisfirePolicy.java (hunk 1 above)
Comment (pegasas) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/MisfirePolicy.java]: Ack. Commited.

**Review by songjianet — APPROVED** [review]

**Pull request closed without merging**